File: front-end/Mobile/src/models/scheduling/ClassSession.js
import BaseModel from '../BaseModel';

export default class ClassSession extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.classSessionId = data.class_session_id || null;
    this.scheduleBlockId = data.schedule_block_id || null;
    this.sessionDate = data.session_date || null;
    this.openedBy = data.opened_by || null;
    this.openedAt = data.opened_at || null;
    this.closedBy = data.closed_by || null;
    this.closedAt = data.closed_at || null;
    this.sessionStatus = data.session_status || 'Open';
  }

  get isOpen() {
    return this.sessionStatus === 'Open';
  }

  get isClosed() {
    return this.sessionStatus === 'Closed';
  }

  get isCancelled() {
    return this.sessionStatus === 'Cancelled';
  }

  static fromApi(data) {
    if (!data) return null;
    return new ClassSession(data);
  }

  toApi() {
    return {
      ...super.toApi(),
      class_session_id: this.classSessionId,
      schedule_block_id: this.scheduleBlockId,
      session_date: this.sessionDate,
      opened_by: this.openedBy,
      opened_at: this.openedAt,
      closed_by: this.closedBy,
      closed_at: this.closedAt,
      session_status: this.sessionStatus,
    };
  }
}

File: back-end/04-ms-scheduling/src/main/java/com/faceattend_edu/scheduling_service/domain/model/ClassSession.java
package com.faceattend_edu.scheduling_service.domain.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.time.LocalDate;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class ClassSession {
    private Long classSessionId;
    private Long scheduleBlockId;
    private LocalDate sessionDate;
    private String sessionStatus;
    private Long openedBy;
    private Instant openedAt;
    private Long closedBy;
    private Instant closedAt;
    private Instant createdAt;
    private Instant updatedAt;
    private Instant deletedAt;
    private UUID createdBy;
    private UUID updatedBy;
    private UUID deletedBy;
    private long rowVersion;

    private static final Set<String> ALLOWED_STATUSES = Set.of("Open", "Closed", "Cancelled");

    public void validate() {
        if (scheduleBlockId == null) throw new IllegalArgumentException("scheduleBlockId is required");
        if (sessionDate == null) throw new IllegalArgumentException("sessionDate is required");
        if (sessionStatus == null || sessionStatus.isBlank()) throw new IllegalArgumentException("sessionStatus is required");
        if (!ALLOWED_STATUSES.contains(sessionStatus)) throw new IllegalArgumentException("sessionStatus must be one of " + ALLOWED_STATUSES);
    }

    public void touchCreated() {
        if (createdAt == null) createdAt = Instant.now();
        if (sessionStatus == null) sessionStatus = "Open";
        if (rowVersion == 0) rowVersion = 1L;
    }

    public void touchUpdated() {
        updatedAt = Instant.now();
        rowVersion++;
    }

    public boolean isOpen() { return "Open".equals(sessionStatus); }
    public boolean isClosed() { return "Closed".equals(sessionStatus); }
    public boolean isCancelled() { return "Cancelled".equals(sessionStatus); }

    @Override public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ClassSession)) return false;
        ClassSession that = (ClassSession) o;
        return Objects.equals(classSessionId, that.classSessionId);
    }
    @Override public int hashCode() { return Objects.hash(classSessionId); }
}
